client/src/worldgen.rs

```diff
@@ -151,39 +153,68 @@ pub fn voxels(graph: &DualGraph, node: NodeId, chunk: Vertex, dimension: u8) ->
                 let center = voxel_center(dimension, coords);
                 let cube_coords = center * 0.5;
 
-                let elev = trilerp(&enviros.max_elevations, cube_coords);
-                let rain = trilerp(&enviros.rainfalls, cube_coords);
-                let slope = trilerp(&enviros.slopeiness, cube_coords);
-
-                let mut voxel_mat = Material::Dirt;
-                let max_e = elev;
-
-                if rain > 2.5 {
+                let rain = triserp(&enviros.rainfalls, cube_coords, 0.2, 0.8);
+                let temp = triserp(&enviros.temperatures, cube_coords, 0.2, 0.8);
+                let slope = triserp(&enviros.slopeinesses, cube_coords, 0.2, 0.8);
+
+                let block = triserp(&enviros.blockinesses, cube_coords, 0.2, 0.8);
+                // map real number block to interval (0, 0.25), biased towards 0
+                let triserp_threshold = 2.0f64.powf(block) / (8.0 + 2.0f64.powf(block)) * 0.25;
+                let elev = triserp(
+                    &enviros.max_elevations,
+                    cube_coords,
+                    triserp_threshold,
+                    1.0 - triserp_threshold,
+                );
```

Comment:
Is the goal of using `triserp`, and introducing `blockiness`, to give the terrain a varying degree of terraced character? If so, we should document that clearly in comments, and retain the simpler and more predictable `trilerp` for the other factors.

Reply:
Additionally, I'm not a huge fan of coupling visible terrain features directly to the topology of the graph, when avoidable. If terraces *are* the goal, would it make more sense to quantize the max elevation value?

Reply:
Yes. Specifically, the point of triserp over trilerp is to give the terrain a terraced character, and the point of introducing blockiness is to make the terracing stronger in some regions than others.

As for coupling visible terrain features to the topology of the graph, I think we just have different opinions of how desirable it is. I am in favour of leaning into it or perhaps even embracing it outright, because it seems really difficult to escape it. If we try to make smooth equidistant surfaces out of voxels, we get pentagonal artifacts. No matter how high the subdivision factor is, there will be places where twenty voxels meet at a corner instead of eight voxels---and all voxels get pointer as they get closer to those corners. Hyperbolic planes aligned with voxel faces crisscross through the space at right angles. That said, my opinion might change after seeing the world with a much higher subdivision factor, and quantizing the max elevation value might yield results better than my approach here.

Reply:
You're absolutely right that completely hiding the topology is impossible, at least so long as we're using discrete cubical voxels, and I agree that we shouldn't try to paper over that. However, I'd still like to minimize the extent to which it influences the world to the cases where it's difficult to avoid. if we're not careful, we'll end up *emphasizing* the topology with highly visible patterns rather than merely rolling with it, which I don't think is desirable; we want terrain features at many scales, with no one scale having too much primacy.

I think quantization would produce the intended effect, and would in fact significantly increase the potential variety of terracing intensities. The implementation would also be simpler and more efficient, since we'd be quantizing the interpolated value rather than using an ad-hoc nonlinear interpolation. Given those advantages, and the absence of any clear drawbacks, I'd like to explore that approach.